### PR TITLE
EPL-1292: How to create a Report

### PR DESCRIPTION
**Pull request by mapenglishservices**

### CODE REVIEW

File: docs/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+  * On Linux: just download the .tgz file and uncompress it, execute the binary _Jaspersoft Studio_ located inside the main folder.
+  * On Windows: Download and execute the .exe file.
+
+###  Configuring Jaspersoft Studio Library Version
+!!! warning
+    Etendo supports Jasper Reports compatible with JasperReports 6.0.0. So, if you are using a newer version of Jaspersoft Studio, you need to make sure that the version of the JasperReports library is 6.0.0. It is very important to use the correct version of JasperReports library,so the jrxml file has the syntax compatible with Etendo.
+
+* Go to `Settings` > `Jaspersoft Studio` > `Compatibility` > `Source .jrxml Version` and set it to 6.0.0.
+
+![](../../../assets/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report-1.png)
```
Comment (isaiasb-etendo) [inline docs/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md:35]: @valenvivaldi  esta imagen debería ser en IntelliJ y no en Eclipse

**Review by isaiasb-etendo — CHANGES_REQUESTED** [review]

File: docs/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+  - create a report
+  - reports
+  - Jaspersoft Studio
+  - classpath
+---
+#  How to create a Report
+  
+##  Overview
+
+In this section, the user can find information about the steps required to create a new report in Etendo Classic. The example explained is a simple report with a list of
```
Comment (valeg-etendo) [inline docs/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md:13]: Delete line break

File: docs/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+products.
+
+##  Setting up Jaspersoft Studio
+  
+!!!note
+    It is recommended to use the latest version of Jaspersoft Studio.  
+  
+  
+First, you need to download Jaspersoft Studio, a graphical tool that allows you to
+create and modify JasperReports templates (.jrxml files).
```
Comment (valeg-etendo) [inline docs/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md:23]: Delete line break

File: docs/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+    Etendo supports Jasper Reports compatible with JasperReports 6.0.0. So, if you are using a newer version of Jaspersoft Studio, you need to make sure that the version of the JasperReports library is 6.0.0. It is very important to use the correct version of JasperReports library,so the jrxml file has the syntax compatible with Etendo.
+
+* Go to `Settings` > `Jaspersoft Studio` > `Compatibility` > `Source .jrxml Version` and set it to 6.0.0.
+
+![](../../../assets/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report-1.png)
+
+###  Configuring Jaspersoft Studio Properties
+
+Some properties of Jaspersoft Studio need to be modified in order to work
+properly. In short, you need to make sure:
```
Comment (valeg-etendo) [inline docs/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md:40]: Delete line break

File: docs/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+  * You modified the JasperReport property _net.sf.jasperreports.awt.ignore.missing.font_ and set it to *true*. It can be changed in `Properties` > `Jaspersoft Studio` > `Properties` 
+  * Do not use any *Scriplet* class 
+  * Use *Java* as default expression language 
+
+
+  
+###  Setting up Classpath
+
+In Jaspersoft Studio, each report is supposed to be part of a project. So, you
+first need to create a new project (`File` > `New` > `Project`).
```
Comment (valeg-etendo) [inline docs/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md:51]: Delete line break

File: docs/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+  * The *New Report Wizard* will be opened 
+
+![](../../../assets/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report-2.png)
+
+  * Select a Report Template, (_Blank_ following our example) 
+  * Define a Report Name 
+  * Define the file Location in the project. 
+
+Later on, we will copy this .jrxml file inside our Etendo module that is
+going to keep our Report and also the required configuration in the Application
```
Comment (valeg-etendo) [inline docs/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md:74]: Delete line break

File: docs/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md [hunk 7]
```diff
@@ -0,0 +1,305 @@
...
+  * Pick *Jasper Report*
+  * The *New Report Wizard* will be opened 
+
+![](../../../assets/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report-2.png)
+
+  * Select a Report Template, (_Blank_ following our example) 
+  * Define a Report Name 
+  * Define the file Location in the project. 
+
+Later on, we will copy this .jrxml file inside our Etendo module that is
```
Comment (valeg-etendo) [inline docs/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md:73]: Delete line break

File: docs/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md [hunk 8]
```diff
@@ -0,0 +1,305 @@
...
+going to keep our Report and also the required configuration in the Application
+Dictionary.
```
Comment (valeg-etendo) [inline docs/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md:75]: Delete line break

File: docs/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md [hunk 9]
```diff
@@ -0,0 +1,305 @@
...
+    * *Username:* tad (you can check your username/password in gradle.properties configuration file. For more information about gradle.properties, visit [Install Etendo](../../../getting-started/installation.md#install-etendo)) 
+    * *Password:* tad 
+  * Click the Finish button to generate the JDBC Connection 
+  * Test your connection 
+  * Save 
+
+![](../../../assets/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report-3.png)
+
+Now we have to configure the query: we are going to list the products present
+in the database.
```
Comment (valeg-etendo) [inline docs/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md:93]: Delete line break

File: docs/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md [hunk 10]
```diff
@@ -0,0 +1,305 @@
...
+!!!note
+    It is recommended to use Dejavu fonts in jasper reports because these
+    fonts support most of the characters in almost all languages​​. Besides,
+    Dejavu typography is the family of fonts that Etendo included in
+    jasperreports-fonts library.   
+  
+###  Etendo Runtime Environment
+
+The standard reports in Etendo (`src/org/openbravo/erpReports`) make use
+of several methods that reside inside the *Etendo Runtime Environment* ,
```
Comment (valeg-etendo) [inline docs/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md:141]: Delete line break

File: docs/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md [hunk 11]
```diff
@@ -0,0 +1,305 @@
...
+which cannot be executed at design time. For this reason, we provide a .jar file that encapsulates the following adapted methods of
```
Comment (valeg-etendo) [inline docs/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md:142]: Delete line break

File: docs/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md [hunk 12]
```diff
@@ -0,0 +1,305 @@
...
+the `org.openbravo.erpCommon.utility.Utility` Class:
```
Comment (valeg-etendo) [inline docs/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md:143]: Delete line break

File: docs/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md [hunk 13]
```diff
@@ -0,0 +1,305 @@
...
+
+  * `public static BufferedImage showImageLogo`: returns a logo image that is already included in the JAR archive 
+  * `public static String applyCountryDateFormat`: always returns the date formatted in this pattern dd-MM-yyyy 
+  * `public static DecimalFormat getCountryNumberFormat`: just returns the same DecimalFormat received as parameter 
+
+For this, to be able to preview the standard Etendo reports from
+*JasperStudio*, it suffices just to import the .jar file into the classpath
```
Comment (valeg-etendo) [inline docs/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md:150]: Delete line break

File: docs/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md [hunk 14]
```diff
@@ -0,0 +1,305 @@
...
+of the project within JasperStudio.
```
Comment (valeg-etendo) [inline docs/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md:151]: Delete line break

File: docs/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md [hunk 15]
```diff
@@ -0,0 +1,305 @@
...
+
+###  Adding images and logos to a report using the ShowImage API
+
+You can use the Image BLOB reference to display an image for a specific
+report, or one of the Company logos in the application.
```
Comment (valeg-etendo) [inline docs/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md:156]: Delete line break

File: docs/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md [hunk 16]
```diff
@@ -0,0 +1,305 @@
...
+You can use the Image BLOB reference to display an image for a specific
+report, or one of the Company logos in the application. 
+
+To do so:
+
+1. Add an image object to your report
+
+2. set the expression class to "java.awt.Image" and the expression image to a call to the ShowImage function of the Utility class (if you want it to display a standard ImageBLOB image reference, that corresponds to a field added to a tab), or to the ShowImageLogo function if you want to display the logo of an Organization or Client.
+
+Images loaded with this method must not have alpha channel. A transparency
```
Comment (valeg-etendo) [inline docs/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md:164]: Delete line break

File: docs/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md [hunk 17]
```diff
@@ -0,0 +1,305 @@
...
+layer is not supported by the function that loads images in Jasper Reports.
+
+  * If you want to use the ShowImage function, you need to make the expression image look like: 
+    * `org.openbravo.erpCommon.utility.Utility.showImage("IMAGEID")` 
+
+IMAGEID needs to be the UUID of the image you want to show. You could set this
```
Comment (valeg-etendo) [inline docs/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md:170]: Delete line break

File: docs/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md [hunk 18]
```diff
@@ -0,0 +1,305 @@
...
+  * If you want to use the ShowImageLogo function to show one of the logos, you have several options. 
+    * This one will show the Company logo at System level: 
+        * `org.openbravo.erpCommon.utility.Utility.showImageLogo("yourcompanylogin")` 
+    * This one will show the Company log at Client level (the client used will be the one the user logged at): 
+        * `org.openbravo.erpCommon.utility.Utility.showImageLogo("yourcompanymenu")` 
+    * This one will show the Company logo at Organization level: 
+        * `org.openbravo.erpCommon.utility.Utility.showImageLogo("yourcompanydoc", "ORGANIZATIONID")`
+
+ORGANIZATIONID needs to be the UUID of the Organization whose log you want to
+show. You could set this value using a Jasper argument. An example could be
```
Comment (valeg-etendo) [inline docs/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md:182]: Delete line break

File: docs/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md [hunk 19]
```diff
@@ -0,0 +1,305 @@
...
+`org.openbravo.erpCommon.utility.Utility.showImageLogo("yourcompanydoc","4387D62C6486481AB3D148442A6AD34E")̣` being `4387D62C6486481AB3D148442A6AD34E`
```
Comment (valeg-etendo) [inline docs/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md:183]: Delete line break

File: docs/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md [hunk 20]
```diff
@@ -0,0 +1,305 @@
...
+the organization ID.
```
Comment (valeg-etendo) [inline docs/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md:184]: Delete line break

File: docs/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md [hunk 21]
```diff
@@ -0,0 +1,305 @@
...
+    * *Search Key:* ETPF_ProductList (It is a best practice to start with your module's [DB_Prefix](../concepts/Modularity_Concepts.md#DB_prefix)) 
+    * *Name:* Product List 
+    * *UI Pattern:* Report (Using JR templates) 
+    * *Data Access Level:* Client/Organization 
+    * *Handler*: use the default `org.openbravo.client.application.report.BaseReportActionHandler`
+
+![](../../../assets/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report-9.png)
+
+We must copy the .jrxml template file generated with Jaspersoft Studio into
+our module. When using Process Definition to generate a Report, templates need
```
Comment (valeg-etendo) [inline docs/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md:208]: Delete line break

File: docs/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md [hunk 22]
```diff
@@ -0,0 +1,305 @@
...
+    * *Module:* Pick your module 
+    * *Search Key:* ETPF_ProductList (It is a best practice to start with your module's [DB_Prefix](../concepts/Modularity_Concepts.md#DB_prefix)) 
+    * *Name:* Product List 
+    * *UI Pattern:* Report (Using JR templates) 
+    * *Data Access Level:* Client/Organization 
+    * *Handler*: use the default `org.openbravo.client.application.report.BaseReportActionHandler`
+
+![](../../../assets/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report-9.png)
+
+We must copy the .jrxml template file generated with Jaspersoft Studio into
```
Comment (valeg-etendo) [inline docs/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md:207]: Delete line break

File: docs/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md [hunk 23]
```diff
@@ -0,0 +1,305 @@
...
+our module. When using Process Definition to generate a Report, templates need
+to be stored in the web folder of the module. In our example, we place it in
```
Comment (valeg-etendo) [inline docs/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md:209]: Delete line break

File: docs/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md [hunk 24]
```diff
@@ -0,0 +1,305 @@
...
+    * *Description:* Description of the action related to the menu entry 
+    * *Action:* Pick `Process Definition` 
+    * *Process Definition:* Pick your Process Definition (Product List) 
+
+![](../../../assets/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report-11.png)
+
+##  Compiling
+
+After you have registered the report and menu entry in the Application
+Dictionary, you need to compile to generate the necessary code.
```
Comment (valeg-etendo) [inline docs/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md:234]: Delete line break

File: docs/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md [hunk 25]
```diff
@@ -0,0 +1,305 @@
...
+
+For information on how JasperReports handles the layout, check [Making HTML,XLS or CSV friendly reports](https://community.jaspersoft.com/knowledgebase/tips-n-tricks/making-html-xls-or-csv-friendly-reports/){target="\_blank"}.
+
+Also, a tutorial with the basics of how to design a report can be found [here](https://community.jaspersoft.com/knowledgebase/tips-n-tricks/designing-report-jaspersoft-studio/){target="\_blank"}.
+
+###  Configuring Cell Type in XLS Reports
+
+By default, Etendo Classic reporting engine exports the XLS data as strings. This is done in order ensure that the exported data can be read after opening the report with the vast majority of spreadsheet applications.
+
+If we want to have a particular format in a cell of our XLS report, and for
```
Comment (valeg-etendo) [inline docs/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md:260]: Delete line break

File: docs/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md [hunk 26]
```diff
@@ -0,0 +1,305 @@
...
+example, display numbers inside a numeric cell, this default configuration can be overridden at template level.
+
+To override this configuration, the following must be done inside the .jrxml
```
Comment (valeg-etendo) [inline docs/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md:263]: Delete line break

File: docs/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md [hunk 27]
```diff
@@ -0,0 +1,305 @@
...
+report template:
+
+  1. Add the *net.sf.jasperreports.export.xls.detect.cell.type* property with true as its value. 
+  2. Add a *pattern* for the text field that will be displayed in the XLS cell. With the `<pattern>` tag, a fixed pattern can be set and with the `<patternExpression>` tag, it is possible to define a dynamic pattern. 
+
+!!!note
+    The decimal and thousands *separators* used for the numeric
```
Comment (valeg-etendo) [inline docs/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md:270]: Delete line break

File: docs/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md [hunk 28]
```diff
@@ -0,0 +1,305 @@
...
+    cells exported in this way, will be those defined inside the spreadsheet
```
Comment (valeg-etendo) [inline docs/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md:271]: Delete line break

File: docs/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md [hunk 29]
```diff
@@ -0,0 +1,305 @@
...
+
+In [this section](How_to_create_a_Report_with_iReport.md#Registering_the_Report_in_Application_Dictionary), you can find an example about how to create a Report in this way.
+
+###  Creating a Reports in old releases
+
+To get information about reporting in old releases, you can go [here](../concepts/.Reports_obsolete.md).
+
+###  Report Compilation
+  
+When printing a report in the application, it is previously compiled at
```
Comment (valeg-etendo) [inline docs/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md:284]: Delete line break

File: docs/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md [hunk 30]
```diff
@@ -0,0 +1,305 @@
...
+runtime. The result of this report compilation is cached if there are no
```
Comment (valeg-etendo) [inline docs/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md:285]: Delete line break

File: docs/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md [hunk 31]
```diff
@@ -0,0 +1,305 @@
...
+modules in _in development_ status.
+
+Besides, it is possible to handle the state of this cache through a JMX
```
Comment (valeg-etendo) [inline docs/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md:288]: Delete line break

File: docs/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md [hunk 32]
```diff
@@ -0,0 +1,305 @@
...
+extension. Thus, this extension allows to:
+
+  * See if the cache is enabled. 
+  * Enable/Disable the cache. 
+  * See the list of reports whose compilation is stored in cache. 
+  * Clear the cache contents. 
+
+###  Barcodes
+
+It is possible to generate barcodes from JasperReports, using barcode4j or
```
Comment (valeg-etendo) [inline docs/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md:298]: Delete line break

File: docs/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md [hunk 33]
```diff
@@ -0,0 +1,305 @@
...
+barbecue libraries. These libraries are included in [Barcode generation in reports](../../../assets/developer-guide/etendo-classic/how-to-guides/org.openbravo.service.reporting.barcode.zip){target="\_blank"} module.
+
+In Platform Features module, there is an example of a report making use of
```
Comment (valeg-etendo) [inline docs/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md:301]: Delete line break

**Review by valeg-etendo — CHANGES_REQUESTED** [review]

File: docs/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md [hunk 34]
```diff
@@ -0,0 +1,305 @@
...
+    * *Process Definition:* Pick your Process Definition (Product List) 
+
+![](../../../assets/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report-11.png)
+
+##  Compiling
+
+After you have registered the report and menu entry in the Application
+Dictionary, you need to compile to generate the necessary code.
+
+    ```bash
```
Comment (isaiasb-etendo) [inline docs/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md:236]: Falta el terminal 

``bash title="Terminal"
./gradlew smartbuild
``

**Review by isaiasb-etendo — CHANGES_REQUESTED** [review]

**Review by isaiasb-etendo — APPROVED** [review]

File: mkdocs.yml [hunk 35]
```diff
@@ -282,7 +282,7 @@ nav:
 #        - Model Object Mapping: developer-guide/etendo-classic/concepts/Model_Object_Mapping.md
         - Alerts: developer-guide/etendo-classic/concepts/Alerts.md
 #        - Processes: developer-guide/etendo-classic/concepts/Processes.md
-#        - How to create a Report: developer-guide/etendo-classic/concepts/How_to_create_a_Report.md
+        - How to create a Report: developer-guide/etendo-classic/concepts/How_to_create_a_Report.md
```
Comment (valeg-etendo) [inline mkdocs.yml:285]:
```suggestion
        - How to Create a Report: developer-guide/etendo-classic/concepts/How_to_create_a_Report.md
```

File: mkdocs.yml [hunk 36]
```diff
@@ -327,7 +327,7 @@ nav:
 #        - How to create a navigation rule: developer-guide/etendo-classic/how-to-guides/How_to_create_a_navigation_rule.md
 #        - How to create a callout that extends from another callout: developer-guide/etendo-classic/how-to-guides/How_to_create_a_callout_that_extends_from_another_callout.md
 #        - How to create a Report with iReport: developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report_with_iReport.md
-#        - How to create a Report: developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md
+        - How to create a Report: developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md
```
Comment (valeg-etendo) [inline mkdocs.yml:330]:
```suggestion
        - How to Create a Report: developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md
```

File: docs/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md [hunk 37]
```diff
@@ -0,0 +1,272 @@
+---
+tags: 
+  - Etendo classic
+  - create a report
+  - reports
+  - Jaspersoft Studio
+  - classpath
+---
+#  How to create a Report
```
Comment (valeg-etendo) [inline docs/developer-guide/etendo-classic/how-to-guides/How_to_create_a_Report.md:9]:
```suggestion
#  How to Create a Report
```

File: mkdocs.yml (hunk 35 above)
Comment (valeg-etendo) [reply mkdocs.yml:285]: @mapenglishservices @isaiasb-etendo Check duplicate page

**Review by valenvivaldi — APPROVED** [review]

**Review by valeg-etendo — APPROVED** [review]

**Pull request merged**